src/TSne.jl

```diff
@@ -281,4 +263,48 @@ function tsne(X::Matrix, ndims::Integer = 2, reduce_dims::Integer = 0,
     return Y
 end
 
+function computeY(X::Matrix, pca_init::Bool, ndims::Integer, verbose::Bool=true)
+    (n, d) = size(X)
+    if !pca_init
+        verbose && info("Starting with random layout...")
+        Y = randn(n, ndims)
+    else
+        verbose && info("Using the first $ndims components of the data PCA as the initial layout...")
+        if reduce_dims >= ndims
+            Y = X[:, 1:ndims] # reuse X PCA
+        else
+            @assert reduce_dims <= 0 # no X PCA
+            Y = pca(X, ndims)
+        end
+    end
+    Y
+end
+
+"Tsne where `X` is a matrix of values and distance is Euclidean"
+function tsne(X::Matrix; ndims::Integer=2, reduce_dims::Integer = 0, verbose::Bool=true,
+              pca_init::Bool=false, kwargs...)
+    # Initialize variables
+    verbose && info("Initial X Shape is $(size(X))")
+    ndims < size(X, 2) || throw(DimensionMismatch("X has fewer dimensions ($(size(X,2))) than ndims=$ndims"))
+    T = eltype(X)
+    X = X * (one(T)/(std(X)::T)) # note that X is copied
+    if 0<reduce_dims<size(X, 2)
+        reduce_dims = max(reduce_dims, ndims)
+        verbose && info("Preprocessing the data using PCA...")
+        X = pca(X, reduce_dims)
+    end
+
+    # @assert false
+    Y = computeY(X, pca_init,ndims, verbose)
+    tsne(Y, pairwisedists(X); ndims=ndims, verbose=verbose, kwargs...)
+end
+
+"Tsne given a vector of arbitrary objects `X` and measure δ: a::T, b::T -> Real"
+function tsne{T}(X::Vector{T}, δ::Function; ndims::Integer=2,
+                 verbose::Bool=true, kwargs...)
+    distances = [δ(x, y) for x in X, y in X]
```

Comment:
I suppose it should be checked that `distances` is non-negative and symmetrical.